### PR TITLE
fix sasquatch remote bug

### PR DESCRIPTION
**Pull request by tjyung**
* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [x] Did you check UI tests on the sample app? They are not executed on CI.
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Checked the remote delegate was in the sdk and fixed sasquatch to use it properly.

## Related PRs or issues

[#AB70263](https://msmobilecenter.visualstudio.com/Mobile-Center/_boards/board/t/Yoann-Team/Backlog%20Items/?workitem=70263)

### CODE REVIEW

File: AppCenterData/AppCenterData/MSData.m [hunk 1]
```diff
@@ -986,7 +986,7 @@ - (void)synchronizeLocalCacheWithCosmosDbWithToken:(MSTokenResult *)token
   id<MSRemoteOperationDelegate> strongDelegate;
   @synchronized(self) {
     strongDelegate = self.remoteOperationDelegate;
-    if (strongDelegate) {
+    if (strongDelegate && [strongDelegate respondsToSelector:@selector(data:didCompleteRemoteOperation:forDocumentMetadata:withError:)]) {
```
Comment (jwallra) [inline AppCenterData/AppCenterData/MSData.m]: This should also add a test to cover the new behavior

File: Sasquatch/SasquatchObjC/AppDelegate.m [hunk 2]
```diff
@@ -268,12 +268,13 @@ - (void)analytics:(MSAnalytics *)analytics didFailSendingEventLog:(MSEventLog *)
...
-    didCompletePendingOperation:(NSString *)operation
-                    forDocument:(MSDocumentWrapper *_Nullable)document
-                      withError:(MSDataError *_Nullable)error {
+    didCompleteRemoteOperation:(NSString *)operation
+           forDocumentMetadata:(MSDocumentMetadata *_Nullable)documentMetadata
+                     withError:(MSDataError *_Nullable)error {
   NSLog(@"Operation processed: %@ ", operation);
-  if (document) {
-    NSLog(@"Document: Partition : %@, document id : %@, eTag : %@ ", document.partition, document.documentId, document.eTag);
+  if (documentMetadata) {
```
Comment (clpolet) [inline Sasquatch/SasquatchObjC/AppDelegate.m]: We should add this to the swift app too in AppDelegate.swift I think it was a miss before.

Comment (tjyung) [reply Sasquatch/SasquatchObjC/AppDelegate.m]: I searched for all reference to the old method name didCompletePendingOperation and theyare all gone.

File: AppCenterData/AppCenterData/MSData.m (hunk 1 above)
Comment (tjyung) [reply AppCenterData/AppCenterData/MSData.m]: added a test

File: AppCenterData/AppCenterDataTests/MSDataTests.m [hunk 3]
```diff
@@ -2715,4 +2715,108 @@ - (void)testRemoteOperationDelegateMethodIsCalled {
...
+    OCMStub([delegateMock data:OCMOCK_ANY didCompleteRemoteOperation:OCMOCK_ANY forDocumentMetadata:OCMOCK_ANY withError:OCMOCK_ANY])
+    .andDo(^(NSInvocation *invocation) {
+        [invocation retainArguments];
+        [invocation getArgument:&returnedStorageDocument atIndex:4];
+        [expectation fulfill];
+    });
+    
+    [self.sut setRemoteOperationDelegate:delegateMock];
+    
+    //test that delegate accepts the right selector
```
Comment (guperrot) [inline AppCenterData/AppCenterDataTests/MSDataTests.m]:
```suggestion
    // Check that the delegate accepts the right selector.
```

File: AppCenterData/AppCenterDataTests/MSDataTests.m [hunk 4]
```diff
@@ -2715,4 +2715,108 @@ - (void)testRemoteOperationDelegateMethodIsCalled {
...
+    assertThatBool([delegateMock respondsToSelector:expectedSelector], isTrue());
+    
+    //test that delegate accepts the incorrect selector
```
Comment (guperrot) [inline AppCenterData/AppCenterDataTests/MSDataTests.m]:
```suggestion
    // Check that the delegate rejects the incorrect selector.
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -13,6 +13,7 @@
 ### App Center Data
 
 * **[Fix]** Reduce retries on Data-related operations to fail fast and avoid the perception of calls "hanging".
+* **[Fix]** Fix an issue where delegate method `data:didCompletePendingOperation:forDocument:withError:` would throw and exception if not implemented.
```
Comment (guperrot) [inline CHANGELOG.md]:
```suggestion
* **[Fix]** Fix an issue where the optional delegate method `data:didCompletePendingOperation:forDocument:withError:` would throw an exception if not implemented (when using `MSData.setRemoteOperationDelegate`).
```

**Pull request merged**